### ISSUE TITLE
printing of `BigFloat` with small precision could use more digits

### ISSUE THREAD

**Issue by rfourquet** [JuliaLang/julia#22719]
I know very little on this topic, but I found the following confusing and can't see a reason why the printing shouldn't be made more exact:
```
julia> x = BigFloat(1.25, 3)
1.2 # I wish this printed 1.25 instead

julia> x == 1.2
false

julia> x == 1.25
true
```

**Comment by rfourquet** [comment]
cc. @pabloferz

**Comment by pabloferz** [comment]
In your example 3 is the precision in bits so that corresponds to `ceil(3 * log10(2)) + 1 == 2` decimal digits. We call MPFR and `1.2` is what it returns accordingly.

Maybe this is an upstream issue with `mpfr_cmp_d` (which `==(::BigFloat, ::Float64)` calls)?

**Comment by rfourquet** [comment]
Again, I'm really not familiar with those questions, but AFAICT `1.25 = 1*2^0+0*2^-1+1*2^-2`, so `1.25` can be represented exactly with 3 bits of precision. It's true that we can get back `1.25` from `1.2` without loss of precision (i.e. `BigFloat(1.2, 3) == 1.25`, I refer to the (your?) comment in mpfr "the number of decimal places needed to read back the number exactly is `ceil(log(10, 2^precision(x))`"), but I believe that `mpfr_cmp_d` is correct. I didn't investigate the printing functions from MPFR, but isn't it possible to feed them a bigger number of decimals?

**Comment by pabloferz** [comment]
> isn't it possible to feed them a bigger number of decimals?

Well, now that you put it like that, I believe we could

```julia
let x = BigFloat(1.25, 3)
    buf = Base.StringVector(10)
    lng = ccall((:mpfr_sprintf,:libmpfr), Int32, (Ptr{UInt8}, Ptr{UInt8}, Ptr{BigFloat}...), buf, "%.2Re", &x)
    String(buf[1:4])
end
```

**Comment by TotalVerb** [comment]
Isn't this a little bit of a slippery slope though... eventually leading to printing all floats exactly?

**Comment by rfourquet** [comment]
I don't know where the limit of the slope should be, but clearly for me printing `1.2` instead of `1.75`, `1.8` instead of `1.75` is *way* before the limit....

**Comment by pabloferz** [comment]
Currently we use the `mpfr_snprintf` function to tells use how many digits it prints at a given precision and in this case it says 7 (`1.2e+00`, but we scrub the `e+00`) so we follow whatever MPFR decides to print.

We could have our own way of deciding how many digits we want and ask MPFR that instead, but we would have to decide on an heuristic for doing that.

**Comment by TotalVerb** [comment]
Maybe a simple solution is to always ask for at least 10 digits (which is 6 excluding the e+00), and then just truncate the trailing zeros if mpfr gives us them.

**Comment by pabloferz** [comment]
Well, something along those lines might work and doesn't sound too hard to implement.

**Comment by StefanKarpinski** [comment]
Too bad MPFR doesn't implement minimal exact printing.

**Comment by yuyichao** [comment]
It does?

```julia
BigFloat(1.25, 3) == BigFloat(1.2, 3)
```

**Comment by yuyichao** [comment]
I think what it doesn't implement (by default?) is exact printing, and it probably shouldn't. I believe that's also what https://github.com/JuliaLang/julia/issues/22719#issuecomment-313956263 is suggesting.

**Comment by StefanKarpinski** [comment]
> It does?

It doesn't:

```jl
julia> BigFloat("0.1")
1.000000000000000000000000000000000000000000000000000000000000000000000000000002e-01
```

Parsing the output at the same precision does round trip to the same value, but it's not the minimal printed form that does so – since in this case that would be `0.1`.

**Comment by yuyichao** [comment]
OK, sure. Those that doesn't seem to be the issue here either. Seems that we actually want to print less digits sometimes if we want to use the minimal exact printing, which is the opposite with what this issue is asking for...

**Comment by TotalVerb** [comment]
I don't think we can satisfy everyone by fudging the number of digits to print. Minimal roundtrippable printing with a minimum number of digits is probably a good compromise.

**Comment by quinnj** [comment]
So is there actually something to do here? Something we can change for BigFloat printing that would make things better? Just throwing an idea out there, we could do something like:
```julia
julia> x = BigFloat(1.25, 3)
1.2 (with 3 bits precision)
```
That would obviously be a non-`:compact` mode display thing, but might help someone who gets confused on this just at the REPL.
